packages/docid/src/commit-id.ts

```diff
@@ -0,0 +1,225 @@
+ */
+function parseCommit(genesis: CID, commit: CID | string | number = null): CID | null {
+  if (!commit) return null;
+
+  const commitCID = parseCID(commit);
+  if (commitCID) {
+    // CID-like
+    if (genesis.equals(commitCID)) {
+      return null;
+    } else {
+      return commitCID;
+    }
+  } else if (String(commit) === '0') {
+    // Zero as number or string
+    return null;
+  } else {
+    throw new Error('Cannot specify commit as a number except to request commit 0 (the genesis commit)');
+  }
+}
+
+/**
+ * Parse CommitID from string representation.
+ *
+ * @param input - string representation of CommitID, be it base36-encoded string or URL.
+ * @see [[CommitID#toString]]
+ * @see [[CommitID#toUrl]]
+ */
+function fromString(input: string): CommitID {
+  const protocolFree = input.replace('ceramic://', '').replace('/ceramic/', '');
+  if (protocolFree.includes('commit')) {
+    const commit = protocolFree.split('?')[1].split('=')[1];
+    const base = protocolFree.split('?')[0];
+    return DocID.fromString(base).travel(commit);
+  } else {
+    return fromBytes(multibase.decode(protocolFree));
+  }
+}
+
+/**
+ * Commit identifier, includes doctype, genesis CID, commit CID.
+ * Encoded as '<multibase-prefix><multicodec-docid><doctype><genesis-cid-bytes><commit-cid-bytes>'
+ */
+export class CommitID implements DocRef {
+  readonly #doctype: number;
+  readonly #cid: CID;
+  readonly #commit: CID | null; // null ≝ genesis commit
+
+  static fromBytes = fromBytes;
+  static fromString = fromString;
+
+  /**
+   * Create a new DocID.
+   *
+   * @param {string|number}      doctype
+   * @param {CID|string}         cid
+   * @param {CID|string}         commit CID. Pass '0', 0, or omit the value as shorthand for the genesis commit.
+   *
+   * @example
+   * new DocID(<docType>, <CID>|<cidStr>)
+   * new DocID(<docType>, <CID>|<cidStr>, <CommitCID>|<CommitCidStr>)
+   */
+  constructor(doctype: string | number, cid: CID | string, commit: CID | string | number = null) {
+    if (!doctype && doctype !== 0) throw new Error('constructor: doctype required');
+    if (!cid) throw new Error('constructor: cid required');
+    this.#doctype = typeof doctype === 'string' ? doctypes.indexByName(doctype) : doctype;
+    this.#cid = typeof cid === 'string' ? new CID(cid) : cid;
+    this.#commit = parseCommit(this.#cid, commit);
+  }
+
+  /**
+   * Construct DocID, no commit information included
+   */
+  get baseID(): DocID {
+    return new DocID(this.#doctype, this.#cid);
+  }
+
+  /**
+   * Doc type code
+   */
+  get type(): number {
+    return this.#doctype;
+  }
+
+  /**
+   * Doc type name
+   */
+  get typeName(): string {
+    return doctypes.nameByIndex(this.#doctype);
+  }
+
+  /**
+   * Genesis CID
+   */
+  get cid(): CID {
+    return this.#cid;
+  }
+
+  /**
+   * Commit CID
+   */
+  get commit(): CID {
+    return this.#commit || this.#cid;
+  }
+
+  /**
+   * Bytes representation
+   */
+  @Memoize()
```

Comment:
What does this do?

Reply:
As the name suggests, it memoizes the getter. No longer need to manually track last-calculated value in a private variable like _bytes.

If called for the first time, the getter calculates the result. If called later, the previously calculated value is returned.

Reply:
This is the smallest memoization library I know. Has no external dependencies.

Reply:
neat!